### PR TITLE
Update Spectrum DnD docs and rename Spectrum DnD hook

### PR DESCRIPTION
**Pull request by LFDanLu**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 🧢 Your Project:

RSP

### CODE REVIEW

File: packages/@react-spectrum/list/docs/ListView.mdx
Comment (LFDanLu) [inline packages/@react-spectrum/list/docs/ListView.mdx:401]: TODO: the docs locally were behaving kinda buggy (drop indicators not showing up/lagging/arrow key nav between drop targets weirdness) but the same examples worked just fine in the storybook when pasted there. Need to investigate

Comment (LFDanLu) [reply packages/@react-spectrum/list/docs/ListView.mdx:401]: Reverted to before the DnD api simplification was merged into main, issue still persists in the docs. Did some more digging and it looks like somehow 2072a5a949768f56d5b2eee50cae667404c6f771 broke the indicators in the docs but not in storybook??

File: packages/@react-spectrum/dnd/src/useDnDHooks.ts
Comment (LFDanLu) [inline packages/@react-spectrum/dnd/src/useDnDHooks.ts:45]: The type expanded weird in the docs so had to move to separate interface

File: packages/@react-spectrum/list/docs/ListView.mdx
Comment (LFDanLu) [inline packages/@react-spectrum/list/docs/ListView.mdx:365]: Will re-comment this stuff out when approved, we can expose these docs when we do the full release

File: packages/@react-types/shared/src/dnd.d.ts
Comment (LFDanLu) [inline packages/@react-types/shared/src/dnd.d.ts:154]: separated the utility handlers of useDnDHooks so I could have them displayed separately in the docs

**Comment by LFDanLu** [comment]
TODO: add images from aria/dnd docs to dnd.mdx

File: packages/@react-types/shared/src/dnd.d.ts
Comment (devongovett) [reply packages/@react-types/shared/src/dnd.d.ts:154]: This is in the shared types package, and this name is pretty general. `DroppableCollectionUtilityOptions`? Still not sure about the name "utility" but can't think of anything better.

File: packages/@react-spectrum/list/docs/ListView.mdx
Comment (devongovett) [inline packages/@react-spectrum/list/docs/ListView.mdx:372]: We may want to make some use of the `begin collapse`/`end collapse` comments to hide some of the duplicate code between these examples, and maybe also `begin highlight`/`end highlight` (added in the aria docs pr) to highlight the differences. If they aren't too large after that, we could consider showing the code by default rather than collapsing.

File: packages/@react-spectrum/list/docs/ListView.mdx
Comment (devongovett) [reply packages/@react-spectrum/list/docs/ListView.mdx:556]: I think the processor could be simplified as described above.

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (snowystinger) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:215]:
```suggestion
To handle the processing of the drop item data, we create an `itemProcessor` utility function that will be used in each of the drop handlers. See the [aria docs](/react-aria/useDroppableCollection.html#drop-data)
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (snowystinger) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:235]:
```suggestion
Next, we add support for insertion drops by providing a `onInsert` handler. This handler uses the relative drop position to insert the dropped
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (snowystinger) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:388]:
```suggestion
Note that `acceptedDragTypes` can still be used alongside `getDropOperation. It acts as a pre-filter that rejects any drag operations that do not contain the drag types defined within `acceptedDragTypes`. We will 
use the same `acceptedDragTypes` from the previous example, eliminating any drag type checks in our `getDropOperation`. If you implement a more complex scenario, such as only allowing on folder drops if
the dragged items are of a specific mime type, you can use the `types` in `getDropOperation` to check the drag types of the current drag operation.
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (snowystinger) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:404]:
```suggestion
Below is what our droppable ListView looks like after combining everything together.
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (snowystinger) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:491]:
```suggestion
`useDnDHooks` then provides you with a set of hooks that you can pass to the component via its `dndHooks` prop, thus enabling drag and drop operations for the component. This approach allows the drag and drop
implementation to be included only when used, keeping bundle sizes small when unused by an application.
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:115]: This is returning an array via `[...keys].map`

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:230]: Yeah, it isn't technically necessary for this example, I'll go ahead and change it. I liked having the `itemProcessor` since it would be easier to expand on if the user wanted to handle directories/files as well or have separate logic for Android Chrome DnD like we have in our stories

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:360]: sounds good, I'll adjust

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:501]: IMO it is better to have it after the prop table so context isn't lost if the user jumps to this section via the table of context

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:508]: I guess I could set up a separate interface that is the `useDnDHook` options with the high level stuff omitted and used that here, unless there is an alternative way to do so via the InterfaceType component?

File: packages/@react-spectrum/dnd/src/useDnDHooks.ts
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/src/useDnDHooks.ts:45]: If it isn't omitted, calls of `dndHook.useDraggableCollectionState` (like in ListView [here](https://github.com/adobe/react-spectrum/blob/4def693b76deb25342ef818750f5698b70718d0a/packages/%40react-spectrum/list/src/ListView.tsx#L146-L151)) will run into a typescript error complaining that `getItems` isn't provided even though it is provided by the `useDnDHooks` options [here](https://github.com/adobe/react-spectrum/blob/bb120eacb4bfb850ad75dda7320702a5f1d77982/packages/%40react-spectrum/dnd/src/useDnDHooks.ts#L87-L90)

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:508]: Hrm, getting the low level drag and drop apis involves making two more interfaces, one for:
```
export interface DraggableProps extends Omit<DraggableCollectionProps, 'preview' | 'getItems'> {
  /**
   * A function that returns the items being dragged. If not specified, we assume that the collection is not draggable.
   * @default () => []
   */
  getItems?: (keys: Set<Key>) => DragItem[]
}
```
and the other for 

```
export interface DroppableCollectionBaseProps {
  /** Handler that is called when a valid drag enters a drop target. */
  onDropEnter?: (e: DroppableCollectionEnterEvent) => void,
...
}
```

I think I'll make have a section for high level drop props, low level drop drops, and then all the props for `useDnDHook`

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:230]: Another reason is to accommodate for typescript since the items in the drop handlers are of type `DropItem[]` which includes Files and Directories. Unfortunately the type won't reflect the filtering that has already occurred in useDroppableCollection

File: packages/@react-spectrum/list/docs/ListView.mdx
Comment (LFDanLu) [inline packages/@react-spectrum/list/docs/ListView.mdx:495]: Examples still felt kinda long so I've kept them collapsed

File: packages/@react-spectrum/dnd/src/useDnDHooks.ts
Comment (devongovett) [reply packages/@react-spectrum/dnd/src/useDnDHooks.ts:45]: Oh, maybe you can do something like this?

```js
useDraggableCollectionState({...props, ...options, getItems: props.getItems});
```

Comment (LFDanLu) [reply packages/@react-spectrum/dnd/src/useDnDHooks.ts:45]: Doesn't seem to work, the core issue is that `useDraggableCollectionState` will have `getItems` as a required prop due to the type, so components that access `dragAndDropHooks.useDraggableCollectionState` will be expected to provide it. However, typescript doesn't understand that `getItems` should only be provided via `useDragAndDrop`'s options and that if `getItems` isn't provided then the `useDraggableCollectionState` hook won't even be called in this specific ListView setup

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (reidbarber) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:462]: Nit, but I think we should use hyphenated `High-level` and `Low-level`

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (reidbarber) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:158]:
```suggestion
[React Aria docs](react-aria/useDraggableCollection.html#drag-data).
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (reidbarber) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:211]:
```suggestion
via [useListData](react-stately/useListData.html). As a reminder, `useListData` is completely optional here and can be replaced by [useAsyncList](react-stately/useAsyncList.html) or any other state management solution.
```

**Review by snowystinger — APPROVED** [review]

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (reidbarber) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:158]: I guess these aren't broken locally or on prod, but they do break on PR builds if included

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (reidbarber) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:353]:
```suggestion
see the [API section](#low-level-drop-handlers-and-options) below and the `useDroppableCollection` [React Aria docs](react-aria/useDroppableCollection.html#low-level-api).
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (reidbarber) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:492]:
```suggestion
Of the various drag-specific options above, `getAllowedDropOperations` may be of particular interest. When the dragged items are dropped on a drop target created using the React Aria drag and drop hooks,
```

**Review by reidbarber — COMMENTED** [review]
Looks good, just some small notes

**Review by ktabors — COMMENTED** [review]
LGTM

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (devongovett) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:243]: I might make this explicit so it's a little clearer what's happening?
```suggestion
    items.map(async item => JSON.parse(await item.getText('adobe-app')))
```

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (devongovett) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:294]: nit: put a newline above here

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (devongovett) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:323]: I'd put these two options at the top (before the events)

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (devongovett) [inline packages/@react-spectrum/dnd/docs/dnd.mdx:429]: I was unable to get dropping directories between items to work.

**Review by reidbarber — APPROVED** [review]

**Review by devongovett — APPROVED** [review]
above can be fixed in followup

File: packages/@react-spectrum/dnd/docs/dnd.mdx
Comment (LFDanLu) [reply packages/@react-spectrum/dnd/docs/dnd.mdx:429]: derp, forgot to import the drag type, thanks for catching

**Pull request merged**
